drivers/acpi/acpi.c

```diff
@@ -0,0 +1,543 @@
+	ACPI_OBJECT_LIST arg_list;
+	ACPI_OBJECT arg[1];
+
+	arg_list.Count = 1;
+	arg_list.Pointer = arg;
+
+	arg[0].Type = ACPI_TYPE_INTEGER;
+	arg[0].Integer.Value = 1;
+
+	status = AcpiEvaluateObject(NULL, "\\_PIC", &arg_list, NULL);
+	if (ACPI_FAILURE(status)) {
+		LOG_WRN("error While executing \\_pic method: %d", status);
+	}
+
+	return status;
+}
+
+static int acpi_get_irq_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+{
+	ACPI_BUFFER rt_buffer;
+	ACPI_NAMESPACE_NODE *node;
+	int status;
+
+	LOG_DBG("%s", bus_name);
+
+	node = acpi_evaluate_method(bus_name, METHOD_NAME__PRT);
+	if (!node) {
+		LOG_ERR("Evaluation failed for given device: %s", bus_name);
+		return -ENODEV;
+	}
+
+	rt_buffer.Pointer = rt_table;
+	rt_buffer.Length = ACPI_DEBUG_BUFFER_SIZE;
+
+	status = AcpiGetIrqRoutingTable(node, &rt_buffer);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("unable to retrieve IRQ Routing Table: %s", bus_name);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int acpi_retrieve_legacy_irq(struct acpi *bus)
+{
+	return acpi_get_irq_table("_SB.PC00", bus->pci_prt_table, sizeof(bus->pci_prt_table));
+}
+
+static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 level, void *ctx,
+					      void **ret_value)
+{
+	ACPI_NAMESPACE_NODE *node;
+	ACPI_BUFFER rt_buffer;
+	struct acpi *bus = (struct acpi *)ctx;
+	struct acpi_dev *child_dev;
+
+	node = ACPI_CAST_PTR(ACPI_NAMESPACE_NODE, obj_handle);
+	char *path_name;
+	int status;
+	ACPI_DEVICE_INFO *dev_info;
+
+	LOG_DBG("%s %p\n", __func__, node);
+
+	/* get device info such as HID, Class ID etc. */
+	status = AcpiGetObjectInfo(obj_handle, &dev_info);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("AcpiGetObjectInfo failed: %s", AcpiFormatException(status));
+		goto exit;
+	}
+
+	if (bus->num_dev >= CONFIG_ACPI_DEV_MAX) {
+		return AE_NO_MEMORY;
+	}
+
+	child_dev = (struct acpi_dev *)&bus->child_dev[bus->num_dev++];
+	child_dev->handle = obj_handle;
+	child_dev->dev_info = dev_info;
+
+	path_name = AcpiNsGetNormalizedPathname(node, TRUE);
+	if (!path_name) {
+		LOG_ERR("No memory for path_name\n");
+		goto exit;
+	} else {
+		LOG_DBG("Device path: %s\n", path_name);
+		child_dev->path = path_name;
+	}
+
+	rt_buffer.Pointer = NULL;
+	rt_buffer.Length = ACPI_ALLOCATE_LOCAL_BUFFER;
+
+	status = AcpiGetCurrentResources(node, &rt_buffer);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("AcpiGetCurrentResources failed: %s", AcpiFormatException(status));
+	} else {
+		child_dev->res_lst = rt_buffer.Pointer;
+	}
+
+exit:
+
+	return 0;
+}
+
+static int acpi_enum_devices(struct acpi *bus)
+{
+	LOG_DBG("");
+
+	AcpiWalkNamespace(ACPI_TYPE_DEVICE, ACPI_ROOT_OBJECT, ACPI_UINT32_MAX,
+			  dev_resource_enum_callback, NULL, bus, NULL);
+
+	return 0;
+}
+
+static int acpi_init(const struct device *dev)
+{
+	int status;
+	struct acpi *bus = dev->data;
+
+	LOG_DBG("");
+
+	/* For debug version only */
+	ACPI_DEBUG_INITIALIZE();
+
+	status = initialize_acpica();
+	if (status != AE_OK) {
```

Comment:
Why to use different methods for error check? `status != AE_OK` vs `ACPI_FAILURE(status)`? Please use one method in the same code. Lets use `ACPI_FAILURE()` and `ACPI_SUCCESS()`